File: admin_panel/lib/models/scout_report.dart
/// Scout Report model - synced with mobile app
/// Status field: draft, submitted, approved, rejected
class ScoutReport {
  final String id;
  final String playerId;
  final String playerName;
  final String playerPosition;
  final int playerAge;
  final String playerTeam;

  // Match Context
  final DateTime matchDate;
  final String rivalTeam;
  final String score;
  final int minutePlayed;
  final String matchType; // Stadium, TV, etc.

  // Physical
  final int physicalRating;
  final String physicalDescription;

  // Technical
  final int technicalRating;
  final String technicalDescription;

  // Tactical
  final int tacticalRating;
  final String tacticalDescription;

  // Mental
  final int mentalRating;
  final String mentalDescription;

  // Overall
  final double overallRating;
  final double potentialRating;

  // SWOT
  final String strengths;
  final String weaknesses;
  final String risks;
  final String recommendedRole;

  // Meta
  final String scoutId;
  final String scoutName;
  final DateTime createdAt;
  final String description;
  final List<String> imageUrls;
  final String status; // draft, submitted, approved, rejected

  const ScoutReport({
    required this.id,
    required this.playerId,
    required this.playerName,
    required this.playerPosition,
    required this.playerAge,
    required this.playerTeam,
    required this.matchDate,
    required this.rivalTeam,
    required this.score,
    required this.minutePlayed,
    required this.matchType,
    required this.physicalRating,
    required this.physicalDescription,
    required this.technicalRating,
    required this.technicalDescription,
    required this.tacticalRating,
    required this.tacticalDescription,
    required this.mentalRating,
    required this.mentalDescription,
    required this.overallRating,
    required this.potentialRating,
    required this.strengths,
    required this.weaknesses,
    required this.risks,
    required this.recommendedRole,
    required this.scoutId,
    required this.scoutName,
    required this.createdAt,
    required this.description,
    required this.imageUrls,
    this.status = 'submitted',
  });

  ScoutReport copyWith({
    String? id,
    String? playerId,
    String? playerName,
    String? playerPosition,
    int? playerAge,
    String? playerTeam,
    DateTime? matchDate,
    String? rivalTeam,
    String? score,
    int? minutePlayed,
    String? matchType,
    int? physicalRating,
    String? physicalDescription,
    int? technicalRating,
    String? technicalDescription,
    int? tacticalRating,
    String? tacticalDescription,
    int? mentalRating,
    String? mentalDescription,
    double? overallRating,
    double? potentialRating,
    String? strengths,
    String? weaknesses,
    String? risks,
    String? recommendedRole,
    String? scoutId,
    String? scoutName,
    DateTime? createdAt,
    String? description,
    List<String>? imageUrls,
    String? status,
  }) {
    return ScoutReport(
      id: id ?? this.id,
      playerId: playerId ?? this.playerId,
      playerName: playerName ?? this.playerName,
      playerPosition: playerPosition ?? this.playerPosition,
      playerAge: playerAge ?? this.playerAge,
      playerTeam: playerTeam ?? this.playerTeam,
      matchDate: matchDate ?? this.matchDate,
      rivalTeam: rivalTeam ?? this.rivalTeam,
      score: score ?? this.score,
      minutePlayed: minutePlayed ?? this.minutePlayed,
      matchType: matchType ?? this.matchType,
      physicalRating: physicalRating ?? this.physicalRating,
      physicalDescription: physicalDescription ?? this.physicalDescription,
      technicalRating: technicalRating ?? this.technicalRating,
      technicalDescription: technicalDescription ?? this.technicalDescription,
      tacticalRating: tacticalRating ?? this.tacticalRating,
      tacticalDescription: tacticalDescription ?? this.tacticalDescription,
      mentalRating: mentalRating ?? this.mentalRating,
      mentalDescription: mentalDescription ?? this.mentalDescription,
      overallRating: overallRating ?? this.overallRating,
      potentialRating: potentialRating ?? this.potentialRating,
      strengths: strengths ?? this.strengths,
      weaknesses: weaknesses ?? this.weaknesses,
      risks: risks ?? this.risks,
      recommendedRole: recommendedRole ?? this.recommendedRole,
      scoutId: scoutId ?? this.scoutId,
      scoutName: scoutName ?? this.scoutName,
      createdAt: createdAt ?? this.createdAt,
      description: description ?? this.description,
      imageUrls: imageUrls ?? this.imageUrls,
      status: status ?? this.status,
    );
  }
}

File: futveri_backend/app/models/scout_report.py
"""
ScoutReport model for player evaluation reports.
Contains physical, technical, tactical, and mental ratings with SWOT analysis.
"""
from datetime import datetime
from typing import Optional
import uuid

from sqlalchemy import DateTime, Float, ForeignKey, Integer, String, Text, func
from sqlalchemy.dialects.postgresql import ARRAY
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.database import Base


class ScoutReport(Base):
    """
    Scout report model for player evaluations.
    
    Status values:
        - draft: Initial state, not submitted
        - submitted: Submitted for review
        - approved: Approved by admin
        - rejected: Rejected by admin
    """
    __tablename__ = "scout_reports"
    
    id: Mapped[str] = mapped_column(
        String(36),
        primary_key=True,
        default=lambda: str(uuid.uuid4()),
    )
    
    # Player Information
    player_id: Mapped[Optional[str]] = mapped_column(
        String(36),
        ForeignKey("players.id", ondelete="SET NULL"),
        nullable=True,
    )
    player_name: Mapped[str] = mapped_column(
        String(100),
        nullable=False,
    )
    player_position: Mapped[str] = mapped_column(
        String(50),
        nullable=False,
    )
    player_age: Mapped[int] = mapped_column(
        Integer,
        nullable=False,
    )
    player_team: Mapped[str] = mapped_column(
        String(100),
        nullable=False,
    )
    player_image_url: Mapped[Optional[str]] = mapped_column(
        String(500),
        nullable=True,
    )
    
    # Match Context
    match_date: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
    )
    rival_team: Mapped[str] = mapped_column(
        String(100),
        nullable=False,
    )
    score: Mapped[str] = mapped_column(
        String(20),
        nullable=False,
    )
    minute_played: Mapped[int] = mapped_column(
        Integer,
        nullable=False,
    )
    match_type: Mapped[str] = mapped_column(
        String(50),
        nullable=False,
    )  # Stadium, TV, Video, etc.
    
    # Physical Rating
    physical_rating: Mapped[int] = mapped_column(
        Integer,
        nullable=False,
    )  # 1-10
    physical_description: Mapped[str] = mapped_column(
        Text,
        nullable=False,
    )
    
    # Technical Rating
    technical_rating: Mapped[int] = mapped_column(
        Integer,
        nullable=False,
    )  # 1-10
    technical_description: Mapped[str] = mapped_column(
        Text,
        nullable=False,
    )
    
    # Tactical Rating
    tactical_rating: Mapped[int] = mapped_column(
        Integer,
        nullable=False,
    )  # 1-10
    tactical_description: Mapped[str] = mapped_column(
        Text,
        nullable=False,
    )
    
    # Mental Rating
    mental_rating: Mapped[int] = mapped_column(
        Integer,
        nullable=False,
    )  # 1-10
    mental_description: Mapped[str] = mapped_column(
        Text,
        nullable=False,
    )
    
    # Overall Ratings
    overall_rating: Mapped[float] = mapped_column(
        Float,
        nullable=False,
    )  # Calculated average
    potential_rating: Mapped[float] = mapped_column(
        Float,
        nullable=False,
    )  # Scout's potential assessment
    
    # SWOT Analysis
    strengths: Mapped[str] = mapped_column(
        Text,
        nullable=False,
    )
    weaknesses: Mapped[str] = mapped_column(
        Text,
        nullable=False,
    )
    risks: Mapped[str] = mapped_column(
        Text,
        nullable=False,
    )
    recommended_role: Mapped[str] = mapped_column(
        String(100),
        nullable=False,
    )
    
    # Additional Information
    description: Mapped[Optional[str]] = mapped_column(
        Text,
        nullable=True,
    )
    notes: Mapped[Optional[str]] = mapped_column(
        Text,
        nullable=True,
    )
    image_urls: Mapped[Optional[list]] = mapped_column(
        ARRAY(String(500)),
        nullable=True,
        default=list,
    )
    
    # Status & Meta
    status: Mapped[str] = mapped_column(
        String(20),
        default="draft",
        nullable=False,
    )  # draft, submitted, approved, rejected
    
    scout_id: Mapped[str] = mapped_column(
        String(36),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
    )
    
    # Admin who approved/rejected
    reviewed_by: Mapped[Optional[str]] = mapped_column(
        String(36),
        ForeignKey("users.id", ondelete="SET NULL"),
        nullable=True,
    )
    reviewed_at: Mapped[Optional[datetime]] = mapped_column(
        DateTime(timezone=True),
        nullable=True,
    )
    rejection_reason: Mapped[Optional[str]] = mapped_column(
        Text,
        nullable=True,
    )
    
    # Timestamps
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        nullable=False,
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False,
    )
    
    # Relationships
    scout = relationship("User", foreign_keys=[scout_id], backref="reports")
    player = relationship("Player", backref="reports")
    reviewer = relationship("User", foreign_keys=[reviewed_by])
    
    def __repr__(self) -> str:
        return f"<ScoutReport {self.player_name} by {self.scout_id}>"
    
    @property
    def is_editable(self) -> bool:
        """Check if report can be edited."""
        return self.status in ("draft", "rejected")
    
    @property
    def average_rating(self) -> float:
        """Calculate average of all ratings."""
        return (
            self.physical_rating +
            self.technical_rating +
            self.tactical_rating +
            self.mental_rating
        ) / 4.0
